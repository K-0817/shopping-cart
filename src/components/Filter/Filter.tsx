import { useProducts } from 'contexts/products-context';

import * as S from './style';

export const availableSizes = ['XS', 'S', 'M', 'ML', 'L', 'XL', 'XXL'];

const Filter = () => {
  const { filters, filterProducts } = useProducts();

  const selectedCheckboxes = new Set(filters);

  const toggleCheckbox = (label: string) => {
    if (selectedCheckboxes.has(label)) {
      selectedCheckboxes.delete(label);
    } else {
      selectedCheckboxes.add(label);
    }

    const filters = Array.from(selectedCheckboxes) as [];

    filterProducts(filters);
  };

  const createCheckbox = (label: string) => (
    <S.Checkbox label={label} handleOnChange={toggleCheckbox} key={label} />
  );

  const createCheckboxes = () => availableSizes.map(createCheckbox);

  return (
    <S.Container>
      <S.Title>Sizes:</S.Title>
      {createCheckboxes()}
    </S.Container>
  );
};

export default Filter;
