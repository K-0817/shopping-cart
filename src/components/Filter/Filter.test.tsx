import { renderWithThemeProvider } from 'utils/test/test-utils';
import { ProductsProvider } from 'contexts/products-context/';

import Filter from '.';
import { availableSizes } from './Filter';

describe('[components] - Filter', () => {
  const setup = () => {
    return renderWithThemeProvider(
      <ProductsProvider>
        <Filter />
      </ProductsProvider>
    );
  };

  test('should render correctly', () => {
    const view = setup();
    expect(view).toMatchSnapshot();
  });

  test('should render every filter size avaliable', () => {
    const { getByText } = setup();
    expect(availableSizes.every((size) => getByText(size))).toBe(true);
  });
});
