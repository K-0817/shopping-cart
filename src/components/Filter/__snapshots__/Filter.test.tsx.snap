// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] - Filter should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Filter__Container-sc-bj2vay-0"
      >
        <h4
          class="Filter__Title-sc-bj2vay-2 cVDNWF"
        >
          Sizes:
        </h4>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="XS"
            />
            <span
              class="checkmark"
            >
              XS
            </span>
          </label>
        </div>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="S"
            />
            <span
              class="checkmark"
            >
              S
            </span>
          </label>
        </div>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="M"
            />
            <span
              class="checkmark"
            >
              M
            </span>
          </label>
        </div>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="ML"
            />
            <span
              class="checkmark"
            >
              ML
            </span>
          </label>
        </div>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="L"
            />
            <span
              class="checkmark"
            >
              L
            </span>
          </label>
        </div>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="XL"
            />
            <span
              class="checkmark"
            >
              XL
            </span>
          </label>
        </div>
        <div
          class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
        >
          <label>
            <input
              data-testid="checkbox"
              type="checkbox"
              value="XXL"
            />
            <span
              class="checkmark"
            >
              XXL
            </span>
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Filter__Container-sc-bj2vay-0"
    >
      <h4
        class="Filter__Title-sc-bj2vay-2 cVDNWF"
      >
        Sizes:
      </h4>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="XS"
          />
          <span
            class="checkmark"
          >
            XS
          </span>
        </label>
      </div>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="S"
          />
          <span
            class="checkmark"
          >
            S
          </span>
        </label>
      </div>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="M"
          />
          <span
            class="checkmark"
          >
            M
          </span>
        </label>
      </div>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="ML"
          />
          <span
            class="checkmark"
          >
            ML
          </span>
        </label>
      </div>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="L"
          />
          <span
            class="checkmark"
          >
            L
          </span>
        </label>
      </div>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="XL"
          />
          <span
            class="checkmark"
          >
            XL
          </span>
        </label>
      </div>
      <div
        class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
      >
        <label>
          <input
            data-testid="checkbox"
            type="checkbox"
            value="XXL"
          />
          <span
            class="checkmark"
          >
            XXL
          </span>
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
